packages/wallet-sdk/src/sign/scw/SCWSigner.ts

```diff
@@ -244,29 +269,106 @@ export class SCWSigner implements Signer {
     const chain = chains?.find((chain) => chain.id === chainId);
     if (!chain) return false;
 
     if (chain !== this.chain) {
       this.chain = chain;
-      this.storage.storeObject(ACTIVE_CHAIN_STORAGE_KEY, chain);
+      SCWStateManager.storeObject(ACTIVE_CHAIN_STORAGE_KEY, chain);
       this.callback?.('chainChanged', hexStringFromNumber(chain.id));
     }
     return true;
   }
+
+  private async addAddress(request: RequestArguments) {
+    const active = SCWStateManager.getItem(ACTIVE_SUB_ACCOUNT_ID_KEY);
+    if (active) {
+      const subAccountMap = SCWStateManager.loadObject<Record<string, unknown>>(SUB_ACCOUNTS_KEY);
+      const subAccount = subAccountMap?.[active];
+      if (subAccount) {
+        return subAccount;
+      }
+    }
+
+    const signer = get(request, 'params[0].signer') as string;
+    // could also check if the app supplied a getAddress sub account signer method?
+    if (!signer) {
+      // Open the popup before constructing the request message.
+      // This is to ensure that the popup is not blocked by some browsers (i.e. Safari)
+      await this.communicator.waitForPopupLoaded?.();
+      // Create the subaccount key pair
+      const { id, keypair, publicKey } = await generateSubAccountKeypair();
+      // send request to popup
+      const response = await this.sendRequestToPopup({
+        ...request,
+        params: [
+          {
+            signer: publicKey,
+          },
+        ],
+      });
+      // Only store the sub account information after the popup has been closed and the
+      // user has confirmed the creation
+      SCWStateManager.setItem(ACTIVE_SUB_ACCOUNT_ID_KEY, publicKey);
+      SCWStateManager.storeObject(SUB_ACCOUNTS_KEY, { [publicKey]: response });
+      idb.setItem(id, { keypair });
+
+      this.subAccount = getSubAccountFromStorage();
+      return response;
+    }
+    // when apps bring their own keys
+    const response = await this.sendRequestToPopup(request);
+
+    // Only store the sub account information after the popup has been closed and the
+    // user has confirmed the creation
+    SCWStateManager.setItem(ACTIVE_SUB_ACCOUNT_ID_KEY, signer.toString());
+    SCWStateManager.storeObject(SUB_ACCOUNTS_KEY, { [signer]: response });
+
+    this.subAccount = getSubAccountFromStorage();
+    return response;
+  }
+
+  private shouldRequestUseSubAccountSigner(request: RequestArguments) {
+    const sender = getSenderFromRequest(request);
+    // if the sender is undefined, it means the application did not provide a sender
+    // in this case, we assume the request is for the active sub account (if present)
+    // if the sender is defined, we check if it is the same as the active sub account
+    // if not, we use the root account signer
+    return (
+      (this.subAccount && sender === undefined) ||
+      (this.subAccount && this.subAccount.address === sender)
+    );
+  }
+
+  private async sendRequestSubAccountSigner(request: RequestArguments) {
```

Comment:
nit: i personally think this name is a bit confusing, wonder if we should do `signRequestWithSubAccountSigner`, or simply `sendRequestToSubAccountSigner`

Reply:
Yes `sendRequestToSubAccountSigner` is better. Will update